File: apps/backend/src/db/review_store.go
package db

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/lucas/oraculo/apps/backend/src/domain"
)

// ReviewStore provides CRUD operations for reviews backed by SQLite.
type ReviewStore struct {
	db *DB
}

// NewReviewStore returns a ReviewStore that uses the given database connection.
func NewReviewStore(db *DB) *ReviewStore {
	return &ReviewStore{db: db}
}

// scanReview reads a Review from a row scanner.
func scanReview(row interface{ Scan(...any) error }) (*domain.Review, error) {
	var (
		r         domain.Review
		createdAt string
	)
	if err := row.Scan(&r.ID, &r.VersionID, &r.VersionType, &r.Verdict, &r.Comment, &createdAt); err != nil {
		return nil, err
	}
	var err error
	if r.CreatedAt, err = time.Parse(sqliteTimeFmt, createdAt); err != nil {
		return nil, fmt.Errorf("parse created_at: %w", err)
	}
	return &r, nil
}

// Create inserts a new review and propagates the verdict to the parent entity's approval_status.
func (s *ReviewStore) Create(versionID int, versionType domain.VersionType, verdict domain.ReviewVerdict, comment string) (*domain.Review, error) {
	tx, err := s.db.conn.Begin()
	if err != nil {
		return nil, fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback()

	res, err := tx.Exec(
		"INSERT INTO reviews (version_id, version_type, verdict, comment) VALUES (?, ?, ?, ?)",
		versionID, versionType, verdict, comment,
	)
	if err != nil {
		return nil, fmt.Errorf("insert review: %w", err)
	}
	id, err := res.LastInsertId()
	if err != nil {
		return nil, fmt.Errorf("last insert id: %w", err)
	}

	// Map review verdict to approval status.
	var status domain.ApprovalStatus
	switch verdict {
	case domain.ReviewApproved:
		status = domain.ApprovalApproved
	case domain.ReviewRejected:
		status = domain.ApprovalRejected
	}

	// Propagate status to parent entity.
	switch versionType {
	case domain.VersionTypeEpic:
		var epicID int
		err = tx.QueryRow("SELECT epic_id FROM epic_versions WHERE id = ?", versionID).Scan(&epicID)
		if err != nil {
			return nil, fmt.Errorf("get epic_id from version: %w", err)
		}
		_, err = tx.Exec(
			"UPDATE epics SET approval_status = ?, updated_at = datetime('now') WHERE id = ?",
			status, epicID,
		)
		if err != nil {
			return nil, fmt.Errorf("propagate epic approval status: %w", err)
		}
	case domain.VersionTypeStory:
		var storyID int
		err = tx.QueryRow("SELECT story_id FROM story_versions WHERE id = ?", versionID).Scan(&storyID)
		if err != nil {
			return nil, fmt.Errorf("get story_id from version: %w", err)
		}
		_, err = tx.Exec(
			"UPDATE stories SET approval_status = ?, updated_at = datetime('now') WHERE id = ?",
			status, storyID,
		)
		if err != nil {
			return nil, fmt.Errorf("propagate story approval status: %w", err)
		}
	}

	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("commit tx: %w", err)
	}

	return s.GetByID(int(id))
}

// GetByID retrieves a review by its ID.
func (s *ReviewStore) GetByID(id int) (*domain.Review, error) {
	row := s.db.conn.QueryRow(
		"SELECT id, version_id, version_type, verdict, comment, created_at FROM reviews WHERE id = ?", id,
	)
	r, err := scanReview(row)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, fmt.Errorf("review %d: %w", id, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("get review: %w", err)
	}
	return r, nil
}

// ListByVersion returns all reviews for a given version ordered by created_at.
func (s *ReviewStore) ListByVersion(versionID int, versionType domain.VersionType) ([]domain.Review, error) {
	rows, err := s.db.conn.Query(
		"SELECT id, version_id, version_type, verdict, comment, created_at FROM reviews WHERE version_id = ? AND version_type = ? ORDER BY created_at",
		versionID, versionType,
	)
	if err != nil {
		return nil, fmt.Errorf("list reviews: %w", err)
	}
	defer rows.Close()

	var reviews []domain.Review
	for rows.Next() {
		r, err := scanReview(rows)
		if err != nil {
			return nil, fmt.Errorf("scan review: %w", err)
		}
		reviews = append(reviews, *r)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate reviews: %w", err)
	}
	return reviews, nil
}

// ListByStory returns all reviews for all versions of a story, ordered by created_at.
func (s *ReviewStore) ListByStory(storyID int) ([]domain.Review, error) {
	query := `
		SELECT r.id, r.version_id, r.version_type, r.verdict, r.comment, r.created_at
		FROM reviews r
		JOIN story_versions sv ON r.version_id = sv.id AND r.version_type = 'story'
		WHERE sv.story_id = ?
		ORDER BY r.created_at
	`
	rows, err := s.db.conn.Query(query, storyID)
	if err != nil {
		return nil, fmt.Errorf("list reviews by story: %w", err)
	}
	defer rows.Close()

	var reviews []domain.Review
	for rows.Next() {
		review, err := scanReview(rows)
		if err != nil {
			return nil, fmt.Errorf("scan review: %w", err)
		}
		reviews = append(reviews, *review)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate reviews: %w", err)
	}
	if reviews == nil {
		reviews = []domain.Review{}
	}
	return reviews, nil
}

File: apps/orchestrator/src/db/stores/review-store.ts
import type { Database } from "bun:sqlite";
import type { Review } from "../../domain/entities.ts";
import type { ReviewVerdict, VersionType, ApprovalStatus } from "../../domain/enums.ts";
import { ErrNotFound } from "../../domain/errors.ts";

export class ReviewStore {
  constructor(private db: Database) {}

  Create(
    versionId: number,
    type_: VersionType,
    verdict: ReviewVerdict,
    comment: string,
  ): Review {
    this.db.run(
      "INSERT INTO reviews (version_id, version_type, verdict, comment) VALUES (?, ?, ?, ?)",
      [versionId, type_, verdict, comment],
    );
    const id = this.db
      .query<{ id: number }, []>("SELECT last_insert_rowid() as id")
      .get()!.id;

    // Map verdict to approval status
    const status: ApprovalStatus = verdict === "approved" ? "approved" : "rejected";

    // Propagate to parent entity
    if (type_ === "epic") {
      const row = this.db
        .query<{ epic_id: number }, [number]>(
          "SELECT epic_id FROM epic_versions WHERE id = ?",
        )
        .get(versionId);
      if (row) {
        this.db.run(
          "UPDATE epics SET approval_status = ?, updated_at = datetime('now') WHERE id = ?",
          [status, row.epic_id],
        );
      }
    } else {
      const row = this.db
        .query<{ story_id: number }, [number]>(
          "SELECT story_id FROM story_versions WHERE id = ?",
        )
        .get(versionId);
      if (row) {
        this.db.run(
          "UPDATE stories SET approval_status = ?, updated_at = datetime('now') WHERE id = ?",
          [status, row.story_id],
        );
      }
    }

    return this.Get(id);
  }

  Get(id: number): Review {
    const row = this.db
      .query<Review, [number]>(
        "SELECT id, version_id, version_type, verdict, comment, created_at FROM reviews WHERE id = ?",
      )
      .get(id);
    if (!row) throw ErrNotFound("review", id);
    return row;
  }

  ListByEpic(epicId: number): Review[] {
    return this.db
      .query<Review, [number]>(
        `SELECT r.id, r.version_id, r.version_type, r.verdict, r.comment, r.created_at
         FROM reviews r
         JOIN epic_versions ev ON r.version_id = ev.id AND r.version_type = 'epic'
         WHERE ev.epic_id = ?
         ORDER BY r.created_at`,
      )
      .all(epicId);
  }

  ListByStory(storyId: number): Review[] {
    return this.db
      .query<Review, [number]>(
        `SELECT r.id, r.version_id, r.version_type, r.verdict, r.comment, r.created_at
         FROM reviews r
         JOIN story_versions sv ON r.version_id = sv.id AND r.version_type = 'story'
         WHERE sv.story_id = ?
         ORDER BY r.created_at`,
      )
      .all(storyId);
  }
}
